src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,70 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+-- XXX We can chain this with another parser to propagate the error and act
+-- accordingly.
+{-# INLINE parseCharUtf8D' #-}
+parseCharUtf8D' :: Monad m => ParserD.Parser m Word8 Char
+parseCharUtf8D' =
+    let A.Array _ ptr _ = utf8d
+    in ParserD.Parser (step' ptr) initial extract
+
+    where
+
+    {-# INLINE prefix #-}
+    prefix = "Streamly.Internal.Data.Stream.parseCharUtf8WithD:"
+
+    {-# INLINE initial #-}
+    initial = return $ ParserD.IPartial UTF8CharDecodeInit
+
+    {-# INLINE extract #-}
+    extract _ = error $ prefix ++ "Not enough input"
```

Comment:
Move `extract` after `step`, we always write these as per the flow, initial, then step, then extract.

Reply:
I like to write initial and extract together as in most cases they are rather small. But yes, keeping the flow is much better and cleaner